ledger/sandbox/src/main/scala/com/digitalasset/platform/store/CompletionFromTransaction.scala

```diff
@@ -3,29 +3,68 @@
 
 package com.digitalasset.platform.store
 
+import java.time.Instant
+
 import com.daml.ledger.participant.state.v1.TransactionId
+import com.digitalasset.api.util.TimestampConversion.fromInstant
 import com.digitalasset.daml.lf.data.Ref
 import com.digitalasset.ledger.{ApplicationId, CommandId}
 import com.digitalasset.ledger.api.domain
-import com.digitalasset.ledger.api.domain.CompletionEvent
-import com.digitalasset.ledger.api.domain.CompletionEvent.{CommandAccepted, CommandRejected}
+import com.digitalasset.ledger.api.domain.RejectionReason
+import com.digitalasset.ledger.api.domain.RejectionReason.{
+  Disputed,
+  Inconsistent,
+  OutOfQuota,
+  PartyNotKnownOnLedger,
+  SubmitterCannotActViaParticipant,
+  TimedOut
+}
+import com.digitalasset.ledger.api.v1.command_completion_service.{
+  Checkpoint,
+  CompletionStreamResponse
+}
+import com.digitalasset.ledger.api.v1.completion.Completion
+import com.digitalasset.ledger.api.v1.ledger_offset.LedgerOffset
 import com.digitalasset.platform.store.dao.LedgerDao
 import com.digitalasset.platform.store.entries.LedgerEntry
+import com.google.rpc.status.Status
+import io.grpc.Status.Code
 
 // Turn a stream of transactions into a stream of completions for a given application and set of parties
 // TODO Remove this when:
 // TODO - the participant can read completions off the index directly AND
 // TODO - the in-memory sandbox is gone
 private[platform] object CompletionFromTransaction {
 
-  private def toApiOffset(offset: LedgerDao#LedgerOffset): domain.LedgerOffset.Absolute =
-    domain.LedgerOffset.Absolute(Ref.LedgerString.assertFromString(offset.toString))
+  private def toApiCheckpoint(
+      recordedAt: Instant,
+      offset: LedgerDao#LedgerOffset): Option[Checkpoint] =
+    Option(
+      Checkpoint(
+        recordTime = Option(fromInstant(recordedAt)),
+        offset = Option(LedgerOffset(LedgerOffset.Value.Absolute(offset.toString)))))
 
-  private def toApiCommandId(commandId: CommandId): domain.CommandId =
-    domain.CommandId(commandId)
-
-  private def toApiTransactionId(transactionId: TransactionId): domain.TransactionId =
-    domain.TransactionId(transactionId)
+  // We _rely_ on the following compiler flags for this to be safe:
+  // * -Xno-patmat-analysis _MUST NOT_ be enabled
+  // * -Xfatal-warnings _MUST_ be enabled
+  private def toErrorCode(rejection: RejectionReason): Int = {
+    import RejectionReason.{
+      Inconsistent,
+      Disputed,
+      PartyNotKnownOnLedger,
+      OutOfQuota,
+      TimedOut,
+      SubmitterCannotActViaParticipant
+    }
+    rejection match {
+      case _: Inconsistent | _: Disputed | _: PartyNotKnownOnLedger =>
+        Code.INVALID_ARGUMENT.value()
```

Comment:
I think you can move the `.value()` call to the call site.

Reply:
Ah, yes, good catch, thanks.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/4675/commits/43dccf73f460eed53af535076097b6b51d1e601c